src/Compilers/CSharp/Portable/Binder/SwitchBinder_BindPatternSwitch.cs

```diff
@@ -1,51 +1,197 @@
+        // Our ultimate intent is to remove the type SwitchBinder and have the fully functional
+        // switch binding implementation present in this class.
+        private readonly SwitchStatementSyntax _switchSyntax;
+        private TypeSymbol _switchGoverningType;
+        private readonly GeneratedLabelSymbol _breakLabel;
+        private readonly bool _isPatternSwitch;
+
+        internal PatternSwitchBinder(Binder next, SwitchStatementSyntax switchSyntax) : base(next, switchSyntax)
+        {
+            _switchSyntax = switchSyntax;
+            _breakLabel = new GeneratedLabelSymbol("break");
+            // PROTOTYPE(typeswitch): until we generate compatible code using the new binder, we only
+            // bind using the new binder if either the feature flag is set, or the syntax requires
+            // we use the new binder.
+            _isPatternSwitch = ((CSharpParseOptions)switchSyntax.SyntaxTree.Options).IsFeatureEnabled(MessageID.IDS_FeaturePatternMatching);
+        }
+
+        internal override BoundStatement BindSwitchExpressionAndSections(SwitchStatementSyntax node, Binder originalBinder, DiagnosticBag diagnostics)
         {
-            var boundSwitchExpression = originalBinder.GetBinder(node.Expression).BindValue(node.Expression, diagnostics, BindValueKind.RValue);
-            // TODO: any constraints on a switch expression must be enforced here. For example,
-            // it must have a type (not be target-typed, lambda, null, etc)
+            Debug.Assert(_switchSyntax.Equals(node));
+
+            // Bind switch expression and set the switch governing type.
+            var localDiagnostics = DiagnosticBag.GetInstance();
+            var boundSwitchExpression = BindSwitchExpressionAndGoverningType(node.Expression, originalBinder, localDiagnostics);
+            diagnostics.AddRangeAndFree(localDiagnostics);
 
             DefaultSwitchLabelSyntax defaultLabel = null;
-            ImmutableArray<BoundPatternSwitchSection> boundPatternSwitchSections = BindPatternSwitchSections(boundSwitchExpression, node.Sections, originalBinder, ref defaultLabel, diagnostics);
+            ImmutableArray<BoundPatternSwitchSection> switchSections = BindPatternSwitchSections(boundSwitchExpression, node.Sections, originalBinder, ref defaultLabel, diagnostics);
+            var locals = GetDeclaredLocalsForScope(node);
+            var functions = GetDeclaredLocalFunctionsForScope(node);
+            return new BoundPatternSwitchStatement(
+                node, boundSwitchExpression,
+                locals, functions, switchSections, this.BreakLabel, this);
+        }
+
+        private SourceLabelSymbol GetDefaultLabel()
+        {
+            // SwitchLabelsMap: Dictionary for the switch case/default labels.
+            // Default label(s) are indexed on a special DefaultKey object.
+
+            return FindMatchingSwitchLabel(s_defaultKey);
+        }
+
+        private static object _nullKey = new object();
```

Comment:
Sure. Will do in the next PR.

---
In reply to: [68624369](https://github.com/dotnet/roslyn/pull/11485#discussion_r68624369) [](ancestors = 68624369)